autonat/autonat-v2.md

```diff
@@ -0,0 +1,255 @@
+`E_BAD_REQUEST`: the peer was unable to decode the request.
+
+`E_REQUEST_REFUSED`: the peer didn't attempt to serve the request because of
+rate limiting, resource limit reached or blacklisting.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+peer that prevented it from completing the request.
+
+Implementations MUST discard responses with status codes they do not understand. 
+
+### Amplification Attack Prevention
+
+![Interaction](autonat-v2-amplification-attack-prevention.svg)
+
+
+When a client asks a server to dial an address that is not the client's observed
+IP address, the server asks the client to send him some non trivial amount of
+bytes as a cost to dial a different IP address. To make amplification attacks
+unattractive, the number of bytes is decided such that it's sufficiently larger
+than a new connection handshake cost.
+
+On receiving a `DialRequest`, the server selects the first address it is capable
+of dialing. If this selected address has a IP different from the client's
+observed IP, the server sends a `DialDataRequest` message with the selected
+address's index(0 based) and `numBytes` set to a sufficiently large value on the
+`/libp2p/autonat/2/dial` stream
+
+Upon receiving a `DialDataRequest` message, the client decides whether to accept
+or reject the cost of dial. If the client rejects the cost, the client resets
+the stream and the `DialRequest` is considered aborted. If the client accepts
+the cost, the client starts transferring `numBytes` bytes to the server. The
+server on receiving `numBytes` bytes proceeds to dial the candidate address. 
+
+## Implementation Suggestions
+
+For any given address, implementations SHOULD do the following
+- periodically recheck reachability status
+- query multiple peers to determine reachability
+
+The suggested heuristic for implementations is to consider an address reachable
+if more than 3 peers report a successful dial and to consider an address
+unreachable if more than 3 peers report unsuccessful dials. 
+
+Implementations are free to use different heuristics than this one
+
+## RPC Messages
+
+All RPC messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec]. 
+
+All RPC messages on stream `/libp2p/autonat/2/dial` are of type `Message`. A
+`DialRequest` message is sent as a `Message` with the `msg` field set to
+`DialRequest`. `DialResponse` and `DialDataRequest` are handled similarly. 
+
+On stream `/libp2p/autonat/2/attempt`, a `DialAttempt` message is sent directly
+
+```proto
+syntax = "proto3";
+
+message Message {
+    oneof msg {
+        DialRequest dialRequest   = 1;
+        DialResponse dialResponse = 2;
+        DialDataRequest dialDataRequest = 3;
+    }
+}
+
+message DialRequest {
+    repeated bytes addrs = 1;
+    fixed64 nonce = 2;
+}
+
+message DialDataRequest {
+    uint32 addrIdx = 1;
```

Comment:
Instead of an index, why not send the address itself? In my eyes just a small complexity reduction. Feel free to ignore. Based on intuition, the additional bytes (index < address) doesn't have a performance impact.

Reply:
I think your suggestion is right. I'll resolve this while taking up the implementation. This will help us avoid sharing the slice between the thread that makes the dial request and the thread that receives the response. Similar effect can be had by sharing the idx to the requesting thread, but this seems simpler to me. I'll resolve this while implementing.